### PR TITLE
Apply patch from ZOOKEEPER-2044 Prevent CancelledKeyException

### PR DESCRIPTION
**Pull request by ericzundel**
Set also STWO-857 where this pages us for global-zk

### CODE REVIEW

**Comment by ericzundel** [comment]
This PR needs to be made against Jason's repo so it ban be merged in.

File: build.xml [hunk 1]
```diff
@@ -167,7 +167,7 @@ xmlns:maven="antlib:org.apache.maven.artifact.ant">
 
     <!-- cppunit property set -->
     <property name="cppunit.lib" value="/usr/lib"/>
-    <property name="cppunit.m4" value="/usr/share/aclocal"/>
+    <property name="cppunit.m4" value="/usr/local/share/aclocal"/>
```
Comment (JasonRosenberg) [inline build.xml]: ? why this change?

**Comment by JasonRosenberg** [comment]
It looks like there are more changes than you've applied, in the patch attached to the jira issue here: https://issues.apache.org/jira/browse/ZOOKEEPER-2044

thoughts?

Also, I'm not sure that the issue in ZOOKEEPER-2044 matches exactly the issue in: https://sentry.sso.global.square/sq/global-zk/issues/9277625/events/19991693/

?

File: zookeeper-2044.patch [hunk 2]
```diff
@@ -0,0 +1,12 @@
+--- src/java/main/org/apache/zookeeper/server/NIOServerCnxn.java	(revision 1716354)
++++ src/java/main/org/apache/zookeeper/server/NIOServerCnxn.java	(working copy)
+@@ -148,7 +148,8 @@
+                 // We check if write interest here because if it is NOT set,
+                 // nothing is queued, so we can try to send the buffer right
+                 // away without waking up the selector
+-                if ((sk.interestOps() & SelectionKey.OP_WRITE) == 0) {
```
Comment (JasonRosenberg) [inline zookeeper-2044.patch]: I'm confused, it looks like this a patch file being checked in?

Comment (JasonRosenberg) [inline zookeeper-2044.patch]: looks like a similar fix would be needed here: https://github.com/ericzundel/zookeeper/blob/a690e0f7a067dc8ea812727b9d91b379720c1394/src/java/main/org/apache/zookeeper/server/NIOServerCnxn.java#L1145

Comment (JasonRosenberg) [inline zookeeper-2044.patch]: actually, looking more at it, it looks like there must be some sort of race condition, from line 1145, because it does check "sk.isValid()" before calling "sk.interestOps".....Seems the better fix would be to catch CancelledKeyException in the places that "interestOps" is called in NIOServerCnxn (might also be good idea to check isValid() before calling it....

Comment (ericzundel) [inline zookeeper-2044.patch]: I downloaded the patch file and saved it.  There were two there but I thought that one superceded the other but maybe we need both.  All of the comments say that this is a false alarm so maybe just catching the exception and suppressing it is the right thing to do for our own patch.

File: build.xml (hunk 1 above)
Comment (ericzundel) [inline build.xml]: It will not compile on my mac without this change.  /usr/share/aclocal doesn't exist and I don't have permissions to make directories there.  brew on a mac installs the automake/autoconf packages so they are in /usr/local/

http://apple.stackexchange.com/questions/208764/cant-write-to-usr-lib
